[docs] Add EAS Observe library integration guide (#47463)
expo/expo: docs/pages/eas/observe/integrations/third-party.mdx

---
title: Integrate a third-party package with EAS Observe
sidebar_title: Third-party
description: Learn how to add an optional EAS Observe integration to a third-party package.
---

import { Step } from '~/ui/components/Step';

> Third-party packages can integrate with EAS Observe to send events that help developers identify performance and usage issues, especially problems that are difficult to detect from application code alone.

Reviewer: **[suggestion]** This opening sentence runs about 29 words and packs two ideas (what the integration does, and why it matters) into one clause, against the "one thought per sentence" guidance (`style-guide.md#write-and-edit-for-clarity`).

```suggestion
Third-party packages can integrate with EAS Observe to send events that help developers identify performance and usage issues. These are often problems that are difficult to detect from application code alone.
```

> Third-party packages can integrate with EAS Observe to send events that help developers identify performance and usage issues. These are often problems that are difficult to detect from application code alone.

Reviewer 2: ⚠️ **correctness** `mdx-components.md#mandatory-usage`

The Expo Router and React Navigation integration pages both open with a `Prerequisites` block, and this page states no requirements at all, so a package author never learns which SDK version has `getIntegrations()` and the `configure` event. Add one covering the minimum SDK version and an app already set up with EAS Observe.

<!-- docs-review -->

Events should describe actionable issues that developers can fix. For example, a package might report:

- An image that is much larger than the device's screen
- A background task that takes too long to complete
- A native resource that loads slowly

<Step label="1">

## Add `expo-observe` as an optional dependency

> Add `expo-observe` as an optional peer dependency so your package continues to work when it is not installed. Also add it as a development dependency for TypeScript types and tests. Do not add it as a required runtime dependency.

Reviewer 2: ```suggestion
Add `expo-observe` as an optional peer dependency so your package continues to work when it is not installed. Also, add it as a development dependency for TypeScript types and tests. Do not add it as a required runtime dependency.
```

```json package.json

> {
>   "peerDependencies": {
>     "expo-observe": ">=56.0.0"
>   },
>   "peerDependenciesMeta": {
>     "expo-observe": {
>       "optional": true
>     }
>   },
>   "devDependencies": {
>     "expo-observe": "^56.0.0"
>   }
> }

Reviewer: **[design]** Fenced code blocks resolve `{{expoSdkVersion}}` / `{{expoSdkMajorVersion}}` at render time specifically so version numbers do not go stale (`components.md#code-block-variables`). The parallel `package.json` example in `docs/pages/config-plugins/development-for-libraries.mdx` already uses this pattern for the same `peerDependencies`/`devDependencies` shape (`"expo": ">={{expoSdkVersion}}"`). Hardcoding `56.0.0` here will read as stale the moment SDK 57 ships.

```suggestion
{
  "peerDependencies": {
    "expo-observe": ">={{expoSdkVersion}}"
  },
  "peerDependenciesMeta": {
    "expo-observe": {
      "optional": true
    }
  },
  "devDependencies": {
    "expo-observe": "^{{expoSdkVersion}}"
  }
}
```

```

> Load the package with `require()` inside a `try...catch` block. Use `typeof import()` to preserve its TypeScript types:

Reviewer 2: ```suggestion
Load the package with `require()` inside a `try/catch` block. Use `typeof import()` to preserve its TypeScript types:
```

```ts observe.ts
let observeModule: typeof import('expo-observe') | undefined;

try {
  observeModule = require('expo-observe') as typeof import('expo-observe');
} catch {
  // The integration remains disabled when expo-observe is not installed.
}
```

</Step>

<Step label="2">

## Declare the integration configuration

Use declaration merging to add your package's integration key to `expo-observe`:

```ts observe.types.ts
export type YourPackageIntegrationConfig = {
  thresholdMs?: number;
};

declare module 'expo-observe' {
  interface ObserveIntegrationsConfig {
    'your-package'?: boolean | YourPackageIntegrationConfig;
  }
}
```

Export this declaration from your package entry point so TypeScript loads it when users import your package.

> Users can then enable the integration with `Observe.configure()`:

Reviewer 2: 💡 **nit** `prose.md#voice-and-tone`

"Users" reads as app users here, and the paragraph means the developers who install your package.

```suggestion
Developers using your package can then enable the integration with `Observe.configure()`:
```

<!-- docs-review -->

```tsx app/_layout.tsx
import { Observe } from 'expo-observe';

Observe.configure({
  integrations: {
    'your-package': true,
  },
});
```

If the integration accepts options, users can pass a configuration object instead of `true`:

```tsx app/_layout.tsx
Observe.configure({
  integrations: {
    'your-package': {
      thresholdMs: 1500,
    },
  },
});
```

</Step>

<Step label="3">

## Initialize the integration

> Read the current configuration with `getIntegrations()`. Then listen for the `configure` event so the app can enable, disable, or update the integration later:

Reviewer 2: ```suggestion
Read the current configuration with `getIntegrations()`. Then, listen for the `configure` event so the app can enable, disable, or update the integration later:
```

    the `configure` event.
  </Requirement>
  <Requirement title="An app already using EAS Observe">
    Follow [Get started](/eas/observe/get-started/) to install `expo-observe` and create your first
    build.
  </Requirement>
</Prerequisites>

<Step label="1">

## Add `expo-observe` as an optional dependency

Add `expo-observe` as an optional peer dependency so your package continues to work when it is not installed. Also, add it as a development dependency for TypeScript types and tests. Do not add it as a required runtime dependency.

```json package.json
{
  "peerDependencies": {
    "expo-observe": ">={{expoSdkVersion}}"
  },
  "peerDependenciesMeta": {
    "expo-observe": {
      "optional": true
    }
  },
  "devDependencies": {
    "expo-observe": "^{{expoSdkVersion}}"
  }
}
```

Load the package with `require()` inside a `try/catch` block. Use `typeof import()` to preserve its TypeScript types:

```ts observe.ts
let observeModule: typeof import('expo-observe') | undefined;

try {
  observeModule = require('expo-observe') as typeof import('expo-observe');
} catch {
  // The integration remains disabled when expo-observe is not installed.

> }

Reviewer 3: Maybe we could expose something like `requireOptionalPackage('expo-observe')` through the `expo` package?

Reviewer: As discussed during the call, I'll keep it this way for now

```

</Step>

<Step label="2">

## Declare the integration configuration

Use declaration merging to add your package's integration key to `expo-observe`:

```ts observe.types.ts
export type YourPackageIntegrationConfig = {
  thresholdMs?: number;
};

declare module 'expo-observe' {
  interface ObserveIntegrationsConfig {
    'your-package'?: boolean | YourPackageIntegrationConfig;
  }
}
```

Export this declaration from your package entry point so TypeScript loads it when users import your package.

Developers using your package can then enable the integration with `Observe.configure()`:

```tsx app/_layout.tsx
import { Observe } from 'expo-observe';

Observe.configure({
  integrations: {
    'your-package': true,
  },
});
```

If the integration accepts options, developers can pass a configuration object instead of `true`:

```tsx app/_layout.tsx
Observe.configure({
  integrations: {
    'your-package': {
      thresholdMs: 1500,
    },
  },
});
```

</Step>

<Step label="3">

## Initialize the integration

> Read the current configuration with `getIntegrations()`. Then, listen for the `configure` event so the app can enable, disable, or update the integration later:

Reviewer 3: Let's consider changing the API for getting the integrations config to something like this 👇 so the integrating library have access to only these integrations it knows about.

```ts
let integrationConfig = Observe.getIntegrationConfig(integrationName)
```

Reviewer: Removed, in favor of `initIntegration`

```ts observe.ts
import type { ObserveIntegrationsConfig } from 'expo-observe';

import { disableObserveIntegration, enableObserveIntegration } from './monitoring';

let initialized = false;

function applyConfiguration(integrations: ObserveIntegrationsConfig) {
  const config = integrations['your-package'];

  if (!config) {
    disableObserveIntegration();
    return;
  }

  enableObserveIntegration(config === true ? {} : config);
}

export function initObserveIntegrationIfNeeded() {
  if (initialized || typeof window === 'undefined' || !observeModule) {
    return;
  }
  initialized = true;

  const { Observe } = observeModule;

  applyConfiguration(Observe.getIntegrations());

>   Observe.addListener('configure', ({ integrations }) => applyConfiguration(integrations));

Reviewer 3: Maybe we can expose a dedicated function that fires the callback immediately if the configuration was already provided and listens for the event otherwise.

```ts
Observe.initIntegration(integrationName, integrationConfig => { ... });
```
